lib/tbot/ssh_proxy.go

```diff
@@ -0,0 +1,262 @@
+			proxy = expanded.Proxy
+		}
+		if expanded.Host != "" {
+			hostPort = expanded.Host
+		}
+	}
+
+	proxyHost, _, err := net.SplitHostPort(proxy)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	facade, keyring, err := parseIdentity(proxyConfig.BotConfig, proxyHost, cluster)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	sshConfig, err := facade.SSHClientConfig()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if user != "" {
+		sshConfig.User = user
+	}
+
+	if cluster == "" {
+		cluster = facade.Get().ClusterName
+	}
+
+	pclt, err := proxyclient.NewClient(ctx, proxyclient.ClientConfig{
+		ProxyAddress:      proxyConfig.ProxyServer,
+		TLSRoutingEnabled: proxyConfig.TLSRoutingEnabled,
+		TLSConfigFunc: func(cluster string) (*tls.Config, error) {
+			cfg, err := facade.TLSConfig()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			// The facade TLS config is tailored toward connections to the Auth service.
+			// Override the server name to be the proxy and blank out the next protos to
+			// avoid hitting the proxy web listener.
+			cfg.ServerName = proxyHost
+			cfg.NextProtos = nil
+			return cfg, nil
+		},
+		UnaryInterceptors:       []grpc.UnaryClientInterceptor{interceptors.GRPCClientUnaryErrorInterceptor},
+		StreamInterceptors:      []grpc.StreamClientInterceptor{interceptors.GRPCClientStreamErrorInterceptor},
+		SSHConfig:               sshConfig,
+		InsecureSkipVerify:      proxyConfig.BotConfig.Insecure,
+		ALPNConnUpgradeRequired: proxyConfig.ConnectionUpgradeRequired,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var target string
+	if expanded == nil || (len(expanded.Search) == 0 && expanded.Query == "") {
+		targetHost, targetPort, err := net.SplitHostPort(hostPort)
+		if err != nil {
+			targetHost = hostPort
+			targetPort = "0"
+		}
+		targetHost = cleanTargetHost(targetHost, proxyHost, cluster)
+		target = net.JoinHostPort(targetHost, targetPort)
+	} else {
+		authConfig, err := pclt.ClientConfig(ctx, cluster)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Override the credentials with the facade, which is tailored for
+		// connections to auth. The proxy client will try to use the TLS
+		// config from above that was explicitly tailored for connecting
+		// to the proxy, and if reused, will result in handshake failures.
+		authConfig.Credentials = []client.Credentials{facade}
+
+		node, err := resolveTargetHost(ctx, authConfig, expanded.Search, expanded.Query)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		proxyConfig.Log.DebugContext(ctx, "found matching SSH host", "host_uuid", node.GetName(), "host_name", node.GetHostname())
+
+		target = net.JoinHostPort(node.GetName(), "0")
+	}
+
+	conn, _, err := pclt.DialHost(ctx, target, cluster, keyring)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if proxyConfig.EnableResumption {
+		conn, err = resumption.WrapSSHClientConn(ctx, conn, func(ctx context.Context, hostID string) (net.Conn, error) {
```

Comment:
I assume that WrapSSHClientConn will manage closing the raw connection from above.